File: Data Structures & Algorithms/interleaving-string/submission-2.cpp
class Solution {
public:
    bool helper(string s1,string s2,string s3,int i,int j){
        if(i + j >= s3.size()){
            return (i >= s1.size() && j>= s2.size());
        }
        
        bool ans = false;
        if(s1[i] == s3[i+j]){
            ans |= helper(s1,s2,s3,i+1,j);
        }
        if(s2[j] == s3[i+j]){
            ans |= helper(s1,s2,s3,i,j+1);
        }
        return ans;
    }

    bool isInterleave(string s1, string s2, string s3) {
        return helper(s1,s2,s3,0,0);
    }
};

File: Data Structures & Algorithms/course-schedule/submission-2.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        """
            must take b before a
            b -> a

            so no cycles

            we could do top sort or graph coloring

            1 <-> 0
            
        """

        adj = defaultdict(list)
        visited = [0 for _ in range(numCourses)]

        for preq in prerequisites:
            adj[preq[1]].append(preq[0])

        def hasCycle(idx: int) -> bool: 
            if visited[idx] == 2:
                return True
            
            ans = False
            visited[idx] = 2

            for neighbor in adj[idx]:
                if hasCycle(neighbor):
                    return True

            visited[idx] = 1
            return ans

        
        for i in range(numCourses):
            if visited[i] == 0:    
                if(hasCycle(i)):
                    return False
            
        print(adj)
        return True
